# [자연어처리]
# 3주차(3-1). 카운트 기반의 문서표현(2)



# 201801544 김진원
* 모든 셀 실행 후 제출하시기 바랍니다.
* 실습 (3-1-1)이 있습니다.


# 한국어 텍스트의 카운트 벡터 변환


# 1. 데이터 살펴보기

In [ ]:
import pandas as pd
df = pd.read_csv('./daum_movie_review.csv')
df.head(10)

review  rating        date  \
0                             돈 들인건 티가 나지만 보는 내내 하품만       1  2018.10.29   
1       몰입할수밖에 없다. 어렵게 생각할 필요없다. 내가 전투에 참여한듯 손에 땀이남.      10  2018.10.26   
2  이전 작품에 비해 더 화려하고 스케일도 커졌지만.... 전국 맛집의 음식들을 한데 ...       8  2018.10.24   
3                                이 정도면 볼만하다고 할 수 있음!       8  2018.10.22   
4                                               재미있다      10  2018.10.20   
5                                           나는 재밌게 봄      10  2018.10.14   
6                                      0.5점은 줄 수 없냐?       0  2018.10.10   
7                     헐..다 죽었어....나중에 앤트맨 보다가도 깜놀...      10  2018.10.08   
8                                              충격 결말       9  2018.10.06   
9                                                응집력       8  2018.10.05   

    title  
0  인피니티 워  
1  인피니티 워  
2  인피니티 워  
3  인피니티 워  
4  인피니티 워  
5  인피니티 워  
6  인피니티 워  
7  인피니티 워  
8  인피니티 워  
9  인피니티 워

In [ ]:
df.shape

(14725, 4)

# 2. 카운트 벡터 생성하기

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
daum_cv = CountVectorizer(max_features=1000)

daum_DTM = daum_cv.fit_transform(df.review) # review를 이용하여 count vector를 학습하고, 변환
print(daum_cv.get_feature_names_out()[:100]) # count vector에 사용된 feature 이름을 반환

['10점' '18' '1987' '1도' '1점' '1점도' '2시간' '2시간이' '2편' '5점' '6점' '7점' '8점'
 'cg' 'cg가' 'cg는' 'cg도' 'cg만' 'good' 'of' 'ㅋㅋ' 'ㅋㅋㅋ' 'ㅋㅋㅋㅋ' 'ㅎㅎ' 'ㅎㅎㅎ'
 'ㅜㅜ' 'ㅠㅠ' 'ㅠㅠㅠ' 'ㅡㅡ' '가는' '가는줄' '가면' '가서' '가슴' '가슴아픈' '가슴이' '가장' '가족'
 '가족과' '가족들과' '가족의' '가족이' '가지고' '간만에' '갈수록' '감독' '감독님' '감독은' '감독의' '감독이'
 '감동' '감동과' '감동도' '감동은' '감동을' '감동이' '감동입니다' '감동적' '감동적이고' '감동적인' '감사드립니다'
 '감사합니다' '감정이' '갑자기' '갔는데' '갔다가' '강철비' '강추' '강추합니다' '같고' '같네요' '같다' '같습니다'
 '같아' '같아요' '같은' '같은데' '같음' '같이' '개연성' '개연성이' '개인적으로' '거의' '겁나' '것도' '것은'
 '것을' '것이' '것이다' '겨울왕국' '결국' '결말' '결말이' '계속' '고맙습니다' '곤지암' '공포' '공포를'
 '공포영화' '관객']


In [ ]:
! pip install konlpy

In [ ]:
from konlpy.tag import Okt # konlpy에서 Twitter 형태소 분석기를 import
okt = Okt()

print('#전체 형태소 결과:', okt.morphs(df.review[1]))
print('#명사만 추출:', okt.nouns(df.review[1]))
print('#품사 태깅 결과', okt.pos(df.review[1]))

#전체 형태소 결과: ['몰입', '할수밖에', '없다', '.', '어렵게', '생각', '할', '필요없다', '.', '내', '가', '전투', '에', '참여', '한', '듯', '손', '에', '땀', '이남', '.']
#명사만 추출: ['몰입', '생각', '내', '전투', '참여', '듯', '손', '땀', '이남']
#품사 태깅 결과 [('몰입', 'Noun'), ('할수밖에', 'Verb'), ('없다', 'Adjective'), ('.', 'Punctuation'), ('어렵게', 'Adjective'), ('생각', 'Noun'), ('할', 'Verb'), ('필요없다', 'Adjective'), ('.', 'Punctuation'), ('내', 'Noun'), ('가', 'Josa'), ('전투', 'Noun'), ('에', 'Josa'), ('참여', 'Noun'), ('한', 'Determiner'), ('듯', 'Noun'), ('손', 'Noun'), ('에', 'Josa'), ('땀', 'Noun'), ('이남', 'Noun'), ('.', 'Punctuation')]


In [ ]:
def my_tokenizer(doc):
    return [token for token, pos in okt.pos(doc) if pos in ['Noun', 'Verb', 'Adjective']]

print("나만의 토크나이저 결과:", my_tokenizer(df.review[1]))

나만의 토크나이저 결과: ['몰입', '할수밖에', '없다', '어렵게', '생각', '할', '필요없다', '내', '전투', '참여', '듯', '손', '땀', '이남']


In [ ]:
# 약간의 시간이 소요됩니다! (약 2분 내외)
from sklearn.feature_extraction.text import CountVectorizer

# 토크나이저와 특성의 최대개수를 지정
daum_cv = CountVectorizer(max_features=1000, tokenizer=my_tokenizer)
# 명사만 추출하고 싶은 경우에는 tokenizer에 'twitter_tag.nouns'를 바로 지정해도 됨

daum_DTM = daum_cv.fit_transform(df.review) # review를 이용하여 count vector를 학습하고, 변환
print(daum_cv.get_feature_names_out()[:100]) # count vector에 사용된 feature 이름을 반환

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


['가' '가는' '가는줄' '가면' '가서' '가슴' '가장' '가족' '가족영화' '가지' '가치' '각색' '간' '간다'
 '간만' '갈' '갈수록' '감' '감독' '감동' '감사' '감사합니다' '감상' '감성' '감정' '감탄' '갑자기' '갔는데'
 '갔다' '갔다가' '강' '강철' '강추' '같고' '같네요' '같다' '같습니다' '같아' '같아요' '같은' '같은데'
 '같음' '개' '개그' '개봉' '개연' '개인' '거' '거기' '거리' '거의' '걱정' '건' '건가' '건지' '걸'
 '겁니다' '것' '게' '겨울왕국' '결론' '결말' '경찰' '경험' '계속' '고' '고맙습니다' '고민' '고생' '곤지암'
 '곳' '공감' '공포' '공포영화' '과' '과거' '관' '관객' '관객수' '관람' '광주' '괜찮은' '교훈' '구성'
 '국내' '국민' '군인' '군함도' '굿' '권선' '귀신' '그' '그것' '그게' '그날' '그냥' '그닥' '그대로'
 '그때' '그래픽']


In [ ]:
print(repr(daum_DTM))
print(110800/(14725*1000))

<14725x1000 sparse matrix of type '<class 'numpy.int64'>'
	with 110800 stored elements in Compressed Sparse Row format>
0.007524617996604414


In [ ]:
for word, count in zip(daum_cv.get_feature_names_out(), daum_DTM[1].toarray()[0]):
    if count > 0:
        print(word, ':', count, end=', ')

내 : 1, 듯 : 1, 몰입 : 1, 생각 : 1, 손 : 1, 없다 : 1, 할 : 1, 

## 2. 카운트 벡터의 활용

### 코사인 유사도(Cosine similarity)

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

print("#두 번째 리뷰:", df.review[1])
start = len(df.review[1]) // 2 # 두 번째 리뷰의 문자수를 확인하고 뒤 절반을 가져오기 위해 중심점을 찾음
source = str(df.review[1][-start:]) # 중심점으로부터 뒤 절반을 가져와서 비교할 문서를 생성
print('# 중심점 이후의 리뷰:', source)

#두 번째 리뷰: 몰입할수밖에 없다. 어렵게 생각할 필요없다. 내가 전투에 참여한듯 손에 땀이남.
# 중심점 이후의 리뷰: 다. 내가 전투에 참여한듯 손에 땀이남.


In [ ]:
source_cv = daum_cv.transform([source]) # 코사인 유사도는 카운트 벡터에 대해 계산하므로 벡터로 변환
# transform은 반드시 리스트나 행렬 형태의 입력을 요구하므로 리스트로 만들어서 입력
print("#변환 값 확인:")
print(source_cv)
print("#대상 특성 행렬의 크기:", source_cv.shape) # 행렬의 크기를 확인, 문서가 하나이므로 (1, 1000)

#변환 값 확인:
  (0, 160)	1
  (0, 233)	1
  (0, 455)	1
#대상 특성 행렬의 크기: (1, 1000)


In [ ]:
sim_result = cosine_similarity(source_cv, daum_DTM) # 변환된 count vector와 기존 값들과의 similarity 계산

print("#유사도 계산 행렬의 크기:", sim_result.shape)
print("#유사도 계산결과를 역순으로 정렬:", sorted(sim_result[0], reverse=True)[:10])

#유사도 계산 행렬의 크기: (1, 14725)
#유사도 계산결과를 역순으로 정렬: [0.6546536707079772, 0.5773502691896258, 0.5773502691896258, 0.5773502691896258, 0.5773502691896258, 0.5773502691896258, 0.5773502691896258, 0.5773502691896258, 0.4629100498862758, 0.4364357804719848]


In [ ]:
import numpy as np
print('#가장 유사한 리뷰의 인덱스:', np.argmax(sim_result[0]))

#가장 유사한 리뷰의 인덱스: 1


In [ ]:
print('#가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result[0]).argsort()[:10])

#가장 유사한 리뷰부터 정렬한 인덱스: [    1 13318  7414  5840  3866  2523  4406    11  2091  8534]


## 3. TF-IDF로 성능을 높여보자

In [ ]:
# 약간의 시간 소요 (약 2분 내외)
from sklearn.feature_extraction.text import TfidfVectorizer

tf = TfidfVectorizer(max_features=1000, tokenizer=my_tokenizer) # TfidfVectorizer(vocabulary=word_features)
reviews_tf = tf.fit_transform(df.review)

source_tf = tf.transform([source]) # 코사인 유사도는 카운트 벡터에 대해 계산하므로 벡터로 변환
# transform은 반드시 리스트나 행렬 형태의 입력을 요구하므로 리스트로 만들어서 입력

sim_result_tf = cosine_similarity(source_tf, reviews_tf) #변환된 count vector와 기존 값들과의 similarity 계산

print('#가장 유사한 리뷰의 인덱스:', np.argmax(sim_result_tf[0]))

#가장 유사한 리뷰의 인덱스: 1


In [ ]:
print(source_tf)

  (0, 455)	0.7148965827761523
  (0, 233)	0.49966630559377534
  (0, 160)	0.4891384865140424


In [ ]:
# 455, 233, 160번 단어 확인
index_to_word = {v: k for k, v in tf.vocabulary_.items()}
print(index_to_word[455])
print(index_to_word[233])
print(index_to_word[160])

손
듯
내


In [ ]:
print('#카운트 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result[0]).argsort()[:10])

#카운트 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스: [    1 13318  7414  5840  3866  2523  4406    11  2091  8534]


In [ ]:
for i in (-sim_result[0]).argsort()[:10]:
    print(df.review[i])

몰입할수밖에 없다. 어렵게 생각할 필요없다. 내가 전투에 참여한듯 손에 땀이남.
탈출인 듯
개취인듯...
개취인듯 ㅠ
옛날 플래쉬 게임 부기맨과 아웃라스트 휠체어 합쳐 놓은 듯 쩝...........
익숙한 듯 아련한.
진짜 무서웠음..ㄷㄷ 얘들이랑 보는데 내 옆에 친구는 무서워서 내 손 잡고 힘을 엄청 쥐는데..손에 피가 안통했음..
내가졸라이상하네
현실적인듯 비현실적이고 비현실적인듯 현실적인  다채로운듯 단순하고 단순한듯 다채로운 영화
내가본 최악의 영화... 좋은뎃글 알바인듯... 같이본 사람든반응...뭐야 이거


In [ ]:
print('#TF-IDF 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result_tf[0]).argsort()[:10])

#TF-IDF 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스: [    1 13339  4406  9270 11546  2523 13318  7414  5840  3866]


In [ ]:
for i in (-sim_result_tf[0]).argsort()[:10]:
    print(df.review[i])

몰입할수밖에 없다. 어렵게 생각할 필요없다. 내가 전투에 참여한듯 손에 땀이남.
정말 손에 땀을쥐고 보았네요~~
진짜 무서웠음..ㄷㄷ 얘들이랑 보는데 내 옆에 친구는 무서워서 내 손 잡고 힘을 엄청 쥐는데..손에 피가 안통했음..
손수건 필히지참하세요 안그러면  손으로닦아야  합니다
이 영화.. 흥행 100만 이상에 내 손모가지 건다.
익숙한 듯 아련한.
탈출인 듯
개취인듯...
개취인듯 ㅠ
옛날 플래쉬 게임 부기맨과 아웃라스트 휠체어 합쳐 놓은 듯 쩝...........


---

# [[실습 3-1-1]] 지난 시간에 배운 NLTK의 movie review 데이터 셋에 대해서 실습해보기
* **movie review dataset의 다섯 번째 리뷰**에 대해, 아래의 결과를 출력하는 코드를 작성해보시오. (그 외는 모두 자유)
```
print('#카운트 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result[0]).argsort()[:10])
for i in (-sim_result[0]).argsort()[:10]:
    print(df.review[i])
print('#TF-IDF 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result_tf[0]).argsort()[:10])
for i in (-sim_result_tf[0]).argsort()[:10]:
    print(df.review[i])
```

* 마크다운을 적절히 사용하여 코드 및 실행 결과에 대해 설명하시오
* 참고: 2-2. 카운트 기반의 문서표현 중 2. 사이킷런을 이용한 카운트 벡터 생성 부분을 실행
* 제출 기한: 9/19(화) 23시 59분 (기존: 9/18(월) -> 변경: 9/19(화))
    * 내일도 수업이 있기도 하고, 난이도가 있다고 생각되어 하루 더 연장합니다. 시간이 더 필요한 경우 하루 정도 더 생각해보고 제출하시기 바랍니다.

In [ ]:
import nltk
from nltk.corpus import movie_reviews
nltk.download('movie_reviews')
reviews = [movie_reviews.raw(fileid) for fileid in movie_reviews.fileids()]
print("#다섯 번째 리뷰:", reviews[4])

#다섯 번째 리뷰: synopsis : a mentally unstable man undergoing psychotherapy saves a boy from a potentially fatal accident and then falls in love with the boy's mother , a fledgling restauranteur . 
unsuccessfully attempting to gain the woman's favor , he takes pictures of her and kills a number of people in his way . 
comments : stalked is yet another in a seemingly endless string of spurned-psychos-getting-their-revenge type movies which are a stable category in the 1990s film industry , both theatrical and direct-to-video . 
their proliferation may be due in part to the fact that they're typically inexpensive to produce ( no special effects , no big name stars ) and serve as vehicles to flash nudity ( allowing them to frequent late-night cable television ) . 
stalked wavers slightly from the norm in one respect : the psycho never actually has an affair ; on the contrary , he's rejected rather quickly ( the psycho typically is an ex-lover , ex-wife , or ex-husband ) . 
other than that , st

[nltk_data] Downloading package movie_reviews to /root/nltk_data...
[nltk_data]   Package movie_reviews is already up-to-date!


* nltk에서 movie_reviews를 다운로드 한 후 reivews 리스트에 모든 리뷰를 저장한 뒤 다섯 번째 리뷰를 출력해준다.

In [ ]:
reviews_cv = CountVectorizer(max_features=1000)
reviews_DTM = reviews_cv.fit_transform(reviews)

source = str(reviews[4])
source_cv = reviews_cv.transform([source])
sim_result = cosine_similarity(source_cv, reviews_DTM)

* 최대 1000개 특징을 사용하는 cv객체와 문서-단어 행렬을 생성한 후 다섯번째 리뷰의 문자열을 source에 저장, 코사인 유사도를 사용하여 다른 리뷰들과의 유사성을 계산한다.

In [ ]:
tf = TfidfVectorizer(max_features=1000) # TfidfVectorizer(vocabulary=word_features)
reviews_tf = tf.fit_transform(reviews)
source_tf = tf.transform([source])
sim_result_tf = cosine_similarity(source_tf, reviews_tf) #변환된 count vector와 기존 값들과의 similarity 계산

* 최대 1000개의 특징을 사용하는 TfidfVectorizer 객체를 생성하고 TF-IDF 문서-단어 행렬(DTM)을 생성한 뒤 5번째 리뷰를 TF-IDF 벡터로 변환하여 저장, TF-IDF를 사용하여 다섯번째 리뷰와 다른 리뷰들과의 유사성을 계산한다.

In [ ]:
print('#카운트 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result[0]).argsort()[:10])
for i in (-sim_result[0]).argsort()[:10]:
  print(reviews[i])
print('#TF-IDF 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스:', (-sim_result_tf[0]).argsort()[:10])
for i in (-sim_result_tf[0]).argsort()[:10]:
  print(reviews[i])

#카운트 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스: [   4 1269  107 1250   27 1601 1368 1935  574 1453]
synopsis : a mentally unstable man undergoing psychotherapy saves a boy from a potentially fatal accident and then falls in love with the boy's mother , a fledgling restauranteur . 
unsuccessfully attempting to gain the woman's favor , he takes pictures of her and kills a number of people in his way . 
comments : stalked is yet another in a seemingly endless string of spurned-psychos-getting-their-revenge type movies which are a stable category in the 1990s film industry , both theatrical and direct-to-video . 
their proliferation may be due in part to the fact that they're typically inexpensive to produce ( no special effects , no big name stars ) and serve as vehicles to flash nudity ( allowing them to frequent late-night cable television ) . 
stalked wavers slightly from the norm in one respect : the psycho never actually has an affair ; on the contrary , he's rejected rather quickly ( the psycho ty

* 코드를 실행한 결과

 카운트 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스: [   4 1269  107 1250   27 1601 1368 1935  574 1453]

 TF-IDF 벡터에 대해 가장 유사한 리뷰부터 정렬한 인덱스: [   4  107  893 1269  100 1060 1837 1862 1250  992]


 * 유사도 측정의 기준이 되는 원본인 5번째 리뷰의 인덱스에 해당하는 4가 각 리스트의 가장 첫번째에 오게 되며 그 후로는 같은 인덱스가 두 리스트 모두에 포함되어 있는 경우도 있지만 유사도의 순서가 다르며 조금 다른 결과가 나오게 되었다.

# [파일] -> [다운로드] -> [.ipynb 다운로드]
# 제출 후 **구글 계정 로그아웃** 잘 하시기 바랍니다.